File: osd/2.1/spock.c
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <signal.h>
#include <sys/wait.h>

void kirks_log(int sig)
{
    int ch;
    char *path = "/tmp/kirk.log";
    FILE *f = fopen(path,"r");
    if(NULL != f)
    {
        while((ch = getc(f)) != EOF)
            putc(ch, stdout);
        fclose(f);
        unlink(path);
    }
    else
        printf("No log found\n");

}

int spock_to_enterprise(void)
{
    char *ext_name = "enterprise";
    //char *ext_name = "nginx";
    int ext_pid;
    FILE *file;
    char *path = "/tmp/enterprise.pid";

    while(1)
    {
        file = fopen(path,"r");
        if(NULL != file)
        {
            fscanf(file,"%d", &ext_pid);
            printf("Spock, %s is at: %d, sending signal...\n", ext_name, ext_pid);
            kill(ext_pid,SIGUSR1);
            fclose(file);
        }
        else
            printf("Spock, %s is down\n", ext_name);
        usleep(random()/30);
    }

    exit(0);
}

int main(void)
{
    pid_t pid;

    char *pidfile = "/tmp/spock.pid";
    FILE *pf = fopen(pidfile,"r");
    if(NULL == pf) 
    {   
        pf = fopen(pidfile,"w");
        fprintf(pf,"%d\n",getpid());
        fclose(pf);
    }   
    else
    {   
        printf("Spock didn't die yet\n");
        exit(1);
    }   


    // Handle SIGUSR1
    signal(SIGUSR1, kirks_log);

    spock_to_enterprise();
    
    unlink(pidfile);
    return 0;
}

File: osd/2.2/spock.py
#!/usr/bin/env python
import signal
import os
import time
import sys

def kirks_log(signum, stack):
    logpath = "/tmp/kirk.log"
    if( os.path.exists(logpath) ):
        f = open(logpath)
        print f.read()
        f.close()
        os.remove(logpath)
    else:
        print "No log found"

def spock_to_enterprise():
    e_pidfile = "/tmp/enterprise.py.pid"
    while True:
        if( os.path.exists(e_pidfile) ):
            f = open( e_pidfile )
            e_pid = f.readline()
            f.close()
            print "Spock, enterprise is at:", e_pid, ", sending signal..."
            os.kill(int(e_pid), signal.SIGUSR1)
            time.sleep(30)
        else:
            print "Spock, enterprise is down"
            time.sleep(30)
    
    sys.exit(0)

def main():
    pidfile = "/tmp/spock.py.pid"
    if( os.path.exists(pidfile) ):
        print "Spock didn't die yet. Exiting..."
        sys.exit(1)
    else:
        f = open(pidfile, "w")
        our_pid = os.getpid()
        f.write(repr(our_pid))
        f.close()
    
    signal.signal(signal.SIGUSR1, kirks_log)
    spock_to_enterprise()
    os.remove(pidfile)
    sys.exit(0)

if __name__ == "__main__":
    main()
